### PR TITLE
Adding OAuth2 key hashing, encryption, openid-connect and out-of-band oauth client docs

### PR DESCRIPTION
**Pull request by tgtshanika**
## Purpose
Updating following doc contents

https://apim.docs.wso2.com/en/latest/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys/
https://apim.docs.wso2.com/en/latest/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys/
https://apim.docs.wso2.com/en/latest/Learn/APISecurity/OAuth2DeepDive/provisioning-out-of-band-oauth-clients/
https://apim.docs.wso2.com/en/latest/Learn/APISecurity/OpenIDConnect/obtaining-user-profile-information-with-openid-connect/

### CODE REVIEW

File: en/docs/Learn/APISecurity/OAuth2DeepDive/provisioning-out-of-band-oauth-clients.md [hunk 1]
```diff
@@ -3,50 +3,74 @@
...
+`https://<Server Host>:9443/devportal   
+      `
+4.  Create an application following the steps given in [Create Application]({{base_path}}/Learn/ConsumeAPI/ManageApplication/create-application).
+    
+5.  Click on the respective application to view the credential details.
+    
+     <a href="../../../../../assets/img/Learn/application-select.png" ><img src="../../../../../assets/img/Learn/application-select.png" alt="App listing" 
+          title="App listing" width="50%" /></a>     
+    
+6.  Click on **Production Keys** section to provision an out-of-band OAuth client for production environment. Notice that you now see a Provide Keys section for your application.
```
Comment (shaniR) [inline en/docs/Learn/APISecurity/OAuth2DeepDive/provisioning-out-of-band-oauth-clients.md]: for the production environment. You should now see a ** Provide Existing OAuth Keys ** section below the **Production Key and Secret** section.

File: en/docs/Learn/APISecurity/OAuth2DeepDive/provisioning-out-of-band-oauth-clients.md [hunk 2]
```diff
@@ -3,50 +3,74 @@
...
-        ![]({{base_path}}/assets/attachments/103334858/103334844.png)
-    2.  Click **Provide Keys** , paste the consumer key and consumer secret pair, which you received in [step 4](#ProvisioningOut-of-BandOAuthClients-step4) , and click **Save** .
-        ![]({{base_path}}/assets/attachments/103334858/103334843.png)
+7.  Paste the consumer key and consumer secret pair, which you received in Step 4 in [Create an external OAuth client](Learn/APISecurity/OAuth2DeepDive/provisioning-out-of-band-oauth-clients/#creating-an-external-oauth-client) , and click **Provide** button.
 
-    You have successfully provisioned an out-of-band OAuth client.
+     <a href="../../../../../assets/img/Learn/update-keys.png" ><img src="../../../../../assets/img/Learn/update-keys.png" alt="update-keys" 
+          title="update-keys" width="50%" /></a>
 
+    Now you have successfully mapped an out-of-band OAuth client for the application. You can test this by subscribing to an API via this application, generating an access token for it and accessing the API.
```
Comment (shaniR) [inline en/docs/Learn/APISecurity/OAuth2DeepDive/provisioning-out-of-band-oauth-clients.md]: OAuth2?

Comment (shaniR) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/provisioning-out-of-band-oauth-clients.md]: Let's point the subscribing and genrating a token related document for steps.

File: en/docs/Learn/APISecurity/OAuth2DeepDive/securing-oauth-token-with-hmac-validation.md [hunk 3]
```diff
@@ -15,14 +15,16 @@ If an attacker uses random tokens to send API requests, API Manager will try to
 ![]({{base_path}}/assets/attachments/103334788/103334789.png)
 #### **WSO2 IS Extension - OAuth Token Generator Extension**
 
-Engage the HMAC OAuth handler in order to do the Keyed-Hash Message Authentication Code (HMAC) validation by adding following into &lt;IS\_HOME&gt;/repository/conf/identity/identity.xml
+Engage the HMAC OAuth handler in order to do the Keyed-Hash Message Authentication Code (HMAC) validation by adding following config inti  &lt;APIM_HOME&gt;/repository/conf/deployment.toml
```
Comment (shaniR) [inline en/docs/Learn/APISecurity/OAuth2DeepDive/securing-oauth-token-with-hmac-validation.md]: Add the following configuration to the  &lt;APIM_HOME&gt;/repository/conf/deployment.toml in order to enable Keyed-Hash Message Authentication Code (HMAC) validation via the HMAC OAuth handler.

File: en/docs/Learn/APISecurity/OAuth2DeepDive/securing-oauth-token-with-hmac-validation.md [hunk 4]
```diff
@@ -15,14 +15,16 @@ If an attacker uses random tokens to send API requests, API Manager will try to
...
-    <IdentityOAuthTokenGenerator>com.sample.lahiru.wso2.hmac.oauth</IdentityOAuthTokenGenerator>
+
+```
+[oauth.extensions]
+token_generator="com.sample.lahiru.wso2.hmac.oauth"
 ```
 
 !!! info
-M ore information on developing OAuth token generator extensions [here](https://docs.wso2.com/display/IS530/Extension+Points+for+OAuth#ExtensionPointsforOAuth-OAuthTokenGenerator) . Code for this particular solution can be found in [oauth-hmac-extension](https://github.com/lahirus/oauth-hmac-extension/blob/master/src/main/java/com/sample/lahiru/wso2/hmac/oauth/OAuthHMACTokenIssuer.java) GitHub repository .
+    More information on developing OAuth token generator extensions [here](https://is.docs.wso2.com/en/5.9.0/learn/extension-points-for-oauth/#oauth-token-generator) . Code for this particular solution can be found in [oauth-hmac-extension](https://github.com/lahirus/oauth-hmac-extension/blob/master/src/main/java/com/sample/lahiru/wso2/hmac/oauth/OAuthHMACTokenIssuer.java) GitHub repository .
```
Comment (shaniR) [inline en/docs/Learn/APISecurity/OAuth2DeepDive/securing-oauth-token-with-hmac-validation.md]: More information on developing OAuth token generator extensions can be found here.

File: en/docs/Learn/APISecurity/OpenIDConnect/obtaining-user-profile-information-with-openid-connect.md [hunk 5]
```diff
@@ -63,28 +61,25 @@ For an online tool to decode the JWT, go to <https://jwt.io/>
...
+{
+  "sub": "user1@carbon.super",
+  "organization": "WSO2",
+  "email": "user1@gmail.com"
+}
 ```
 
 !!! note
-By default, only the username (sub) information will be available in the response. You can customize the user information returned by configuring the claims of the relevant Service Provider generated for the Application created in Store. For more information, see [Service Provider Claim Configuration](https://docs.wso2.com/display/IS530/Configuring+Claims+for+a+Service+Provider) .
+    By default, only the username (sub) information will be available in the response. You can customize the user information returned by configuring the claims of the relevant Service Provider generated for the Application created in Store. For more information, see [Service Provider Claim Configuration](https://is.docs.wso2.com/en/5.9.0/learn/configuring-claims-for-a-service-provider/) .
```
Comment (shaniR) [inline en/docs/Learn/APISecurity/OpenIDConnect/obtaining-user-profile-information-with-openid-connect.md]: do you customize it in the store? or is the service provider generated in the store?

File: en/docs/Learn/APISecurity/OAuth2DeepDive/AccessTokenTypes/opaque-tokens.md [hunk 6]
```diff
@@ -37,7 +37,7 @@ Sign in to the Developer Portal.
 
      <a href="../../../../../assets/img/Learn/create-oauth-application.png" ><img src="../../../../../assets/img/Learn/create-oauth-application.png" alt="Add oauth application" title="Add oauth application" width="70%" /></a>
 
-### Step 3 - Generate a OAuth2 Opaque token
+### Step 3 - Generate a OAuth2 Reference token
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/AccessTokenTypes/opaque-tokens.md]: Fixed

File: en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md [hunk 7]
```diff
@@ -1,26 +1,35 @@
-# Encrypting OAuth Keys
+# Encrypting OAuth2 Tokens 
 
-WSO2 API Manager allows you to encrypt any sensitive OAuth2.0 keys that are created. The API Manager encrypts access tokens, client secrets and authorization codes (this can be extended to any other OAuth2.0 keys if needed) using the primary keystore. The result is encoded in Base64 and stored in the database. The RSA algorithm is used by default and the key strength (1024, 2048, etc) is based on the private key strength of the primary keystore. If SymmetricEncryption is enabled, the API Manager uses the AES algorithm by default, or the algorithm specified for the `SymmetricEncryption.Algorithm` in the `carbon.xml` file.
+WSO2 API Manager facilitates OAuth2 token encryption in order to protect OAuth2 access tokens, refresh tokens, consumer secrets, and authorization codes (this can be extended to any other OAuth2.0 keys if needed) using the primary keystore. The result is encoded in Base64 and stored in the database.
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md]: Done

File: en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md [hunk 8]
```diff
@@ -1,26 +1,35 @@
...
+WSO2 API Manager facilitates OAuth2 token encryption in order to protect OAuth2 access tokens, refresh tokens, consumer secrets, and authorization codes (this can be extended to any other OAuth2.0 keys if needed) using the primary keystore. The result is encoded in Base64 and stored in the database. 
 
 !!! info
         **Symmetric Encryption** is a form of encryption where the same key is used to encrypt and decrypt the message along with a mathematical algorithm. As long as both sender and recipient know the secret key, they can encrypt and decrypt all messages that use this key.
 
 !!! warning
         It is recommended to switch this configuration on/off **before any keys have been generated in your system** . Once token encryption is switched on, the system encrypts all sensitive OAuth2.0 data such as Access Tokens, Consumer Secrets, etc. When reading that information, the system assumes that they are in the encrypted format and attempts to decrypt them. Therefore, switching this configuration on **after** any keys are created would break the system, unless the data is converted back into plain text.
 
+Please follow below steps to enable OAuth token encryption.
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md]: Fixed

File: en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md [hunk 9]
```diff
@@ -1,26 +1,35 @@
...
         It is recommended to switch this configuration on/off **before any keys have been generated in your system** . Once token encryption is switched on, the system encrypts all sensitive OAuth2.0 data such as Access Tokens, Consumer Secrets, etc. When reading that information, the system assumes that they are in the encrypted format and attempts to decrypt them. Therefore, switching this configuration on **after** any keys are created would break the system, unless the data is converted back into plain text.
 
+Please follow below steps to enable OAuth token encryption. 
 
-In order to encrypt the OAuth keys, change the following configurations.
+1.  Open the `<API-M_HOME>/repository/conf/deployment.toml` file and add following config.   
 
-1.  In the `<APIM_HOME>/repository/conf/deployment.toml` file, add following configurations:
+    ``` 
+    [apim.oauth_config]
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md]: Fixed

File: en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md [hunk 10]
```diff
@@ -1,26 +1,35 @@
...
+    !!! Note
+        By default, WSO2 API Manager uses `RSA/ECB/OAEPwithSHA1andMGF1Padding` algorithm for token encryption. If you want to change the algorithm, please add following config to deployment.toml, specifying the preferred algorithm.
+    
+        ```
+        [system.parameter]
+        "org.wso2.CipherTransformation" = "<Algorithm>"
+        ```
+    
+2.  Restart the server. 
+3.  Please follow the steps given in [Generate Application Keys]({{base_path}}/Learn/ConsumeAPI/ManageApplication/GenerateKeys/generate-api-keys) to create a new application, generate application consumer key secrets and obtain an access token
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md]: Fixed

File: en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md [hunk 11]
```diff
@@ -1,26 +1,35 @@
...
+
+    !!! Note
+        By default, WSO2 API Manager uses `RSA/ECB/OAEPwithSHA1andMGF1Padding` algorithm for token encryption. If you want to change the algorithm, please add following config to deployment.toml, specifying the preferred algorithm.
+    
+        ```
+        [system.parameter]
+        "org.wso2.CipherTransformation" = "<Algorithm>"
+        ```
+    
+2.  Restart the server.
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md]: Instructions are given to stop the server if it is running. Relevant doc is pointed to see the command to restart the server.

File: en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md [hunk 12]
```diff
@@ -1,26 +1,35 @@
...
+    
+        ```
+        [system.parameter]
+        "org.wso2.CipherTransformation" = "<Algorithm>"
+        ```
+    
+2.  Restart the server. 
+3.  Please follow the steps given in [Generate Application Keys]({{base_path}}/Learn/ConsumeAPI/ManageApplication/GenerateKeys/generate-api-keys) to create a new application, generate application consumer key secrets and obtain an access token
+
+ Once the token encryption is enabled, all the OAuth2 access tokens, refresh tokens, consumer secrets, and authorization codes will be encrypted in the database.
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/encrypting-oauth-keys.md]: removed

File: en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md [hunk 13]
```diff
@@ -1,64 +1,47 @@
-# Hashing OAuth Keys
+# Hashing OAuth2 Access Tokens
 
-!!! note
-If you want to enable this feature, we strongly recommend to use a fresh database. You will need data migration, if you want to go with the existing database. If you are interested in this feature, please [contact us](https://wso2.com/contact/) .
 
+WSO2 API Manager allows you to enable OAuth2 token hashing to protect OAuth2 access tokens, refresh tokens, consumer secrets, and authorization codes.
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md]: Fixed

File: en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md [hunk 14]
```diff
@@ -1,64 +1,47 @@
...
 
-WSO2 API Manager allows you to hash any sensitive OAuth2.0 keys you create. API Manager can hash access tokens, refresh tokens, client secrets, and authorization codes. Follow the steps below to configure this feature.
+Follow the instructions below to set up OAuth token hashing.
 
-1.  Open the `<API-M_HOME>/repository/conf/api-manager.xml` file. To enable token hashing, set the `<EnableTokenHashMode>` parameter to `true` under `<OAuthConfigurations>.         `
-2.  Open the `<APIM_HOME>/repository/conf/identity/identity.xml` file and add the following configurations. Within the `<OAuth>` root tag, add the `<TokenPersistenceProcessor>` parameter. Add the `<EnableClientSecretHash>` parameter and set it to `true` .
+1.  Open the `<API-M_HOME>/repository/conf/deployment.toml` file and add following config.
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md]: Fixed

File: en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md [hunk 15]
```diff
@@ -1,64 +1,47 @@
...
-3.  Run the following commands against the database.
-    -   Remove the CONN\_APP\_KEY constraint from IDN\_OAUTH2\_ACCESS\_TOKEN table.
-        For a given set of consumer key, user, and scope values, there can be only one active access token. The CON\_APP\_KEY constraint in the IDN\_OAUTH2\_ACCESS\_TOKEN table enforces this by allowing only one active access token for a given set of consumer key, user, and scope values.
-        But with this hashing feature for every access token request a new access token will be issued. So for a given set of consumer key, user, and scope values, there can be multiple ACTIVE access tokens. To keep multiple ACTIVE access tokens, we need to remove the CONN\_APP\_KEY constraint from IDN\_OAUTH2\_ACCESS\_TOKEN table.
-        E.g., For an H2 database
-
-        ``` java
-                ALTER TABLE IDN_OAUTH2_ACCESS_TOKEN DROP CONSTRAINT IF EXISTS CON_APP_KEY
-        ```
+2.  Run the appropriate database command to remove the
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md]: Fixed

File: en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md [hunk 16]
```diff
@@ -1,64 +1,47 @@
...
+            `           CON_APP_KEY          ` constraint in the
+            `           IDN_OAUTH2_ACCESS_TOKEN          ` table enforces this
+            by allowing only one active access token to exist for specified
+            consumer key, user, and scope values.  
+            With regard to hashing, a new access token is issued for every
+            access token request . Therefore, for a given consumer key, user,
+            and scope, there can be multiple active access tokens. To allow
+            existence of multiple active access tokens, you need to remove the
+            `           CONN_APP_KEY          ` constraint from
+            `           IDN_OAUTH2_ACCESS_TOKEN          ` table.
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md]: Fixed

File: en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md [hunk 17]
```diff
@@ -1,64 +1,47 @@
...
-
-        ``` java
-                ALTER TABLE IDN_OAUTH2_ACCESS_TOKEN DROP CONSTRAINT IF EXISTS CON_APP_KEY
-        ```
+2.  Run the appropriate database command to remove the
+        `           CONN_APP_KEY          ` constraint from the
+        `           IDN_OAUTH2_ACCESS_TOKEN          ` table. For example,
+        if you are using an H2 database, you need to run the following
+        command:
+    ``` sql
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md]: Will update the scripts

File: en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md [hunk 18]
```diff
@@ -1,64 +1,47 @@
...
-    ![]({{base_path}}/assets/attachments/103334938/103334934.png)
-8.  The consumer secret and access token will be displayed in a pop-up window as shown below. Make sure you note these down, as this will be displayed only once.
-    ![]({{base_path}}/assets/attachments/103334938/103334933.png)
-9.  The consumer key will be accessible through the UI.
-    ![]({{base_path}}/assets/attachments/103334938/103334932.png)
+3.  Restart the server. 
 
-!!! info
-Regenerating the consumer secret
+4.  Please follow the steps given in [Generate Application Keys]({{base_path}}/Learn/ConsumeAPI/ManageApplication/GenerateKeys/generate-api-keys) to create a new application, generate application consumer key secrets and obtain an access token
```
[the remark this answers is not in the document]
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OAuth2DeepDive/hashing-oauth-keys.md]: Fixed

File: en/docs/Learn/APISecurity/OpenIDConnect/obtaining-user-profile-information-with-openid-connect.md (hunk 5 above)
Comment (tgtshanika) [reply en/docs/Learn/APISecurity/OpenIDConnect/obtaining-user-profile-information-with-openid-connect.md]: Added information to locate the relevant service provider via management console

**Pull request merged**